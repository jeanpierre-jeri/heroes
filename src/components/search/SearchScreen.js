import {useForm} from '../../hooks/useForm'
import { getHeroesByName } from '../../selectors/getHeroesByName'
import HeroCard from '../hero/HeroCard'
import queryString from 'query-string'
import { useLocation, useNavigate } from 'react-router-dom'
import { useMemo } from 'react'

const SearchScreen = () => {

  const location = useLocation()
  const navigate = useNavigate()

  const {q = ''} = queryString.parse(location.search)


  const [{searchText}, handleInputChange] = useForm({searchText: q})
  
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])


  const handleSubmit = (e) => {
    e.preventDefault()
    navigate(`?q=${searchText}`)
  }


  return (
    <>
      <h1 className="mt-5">Buscar Héroe</h1> 
      <hr />

      <div className="row">
        <div className="col-5">

          <form onSubmit={handleSubmit}>
            <div className="mb-3">
              <label htmlFor="search" className="form-label">Buscar</label>
              <input type="text" className="form-control" name='searchText' placeholder="Buscar un héroe..." autoComplete="off" onChange={handleInputChange} value={searchText} />
            </div>

            <button type="submit" className="btn btn-primary">Submit</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          {
            q === ''
            ? <div className='alert alert-info'>Buscar un héroe</div>
            : heroesFiltered.length === 0 && <div className='alert alert-danger'>No se encontraron resultador para: {q}</div>

          }

          {
            heroesFiltered.map(hero => (
              <HeroCard key={hero.id} {...hero} />
            ))
          }
        </div>
      </div>
    </>
  )
}

export default SearchScreen
