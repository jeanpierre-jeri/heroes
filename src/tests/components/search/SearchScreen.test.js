import SearchScreen from '../../../components/search/SearchScreen'

import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

const mockNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

describe('Pruebas en <SearchScreen />', () => {
  test('Debe de mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un hÃ©roe')
  })

  test('debe de mostrar a batman y el input con el valor del query string', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('input').prop('value')).toBe('batman')
  })

  test('debe de mostrar un error si no se encuentra el heroe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    )

    expect(wrapper.find('.alert-danger').text().trim()).toBe(
      'No se encontraron resultador para: batman123'
    )
  })

  test('debe de llamar el navigate a la nueva pantalla', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    )

    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'batman',
      },
    })

    wrapper.find('form').prop('onSubmit')({
      preventDefault() {},
    })

    expect(mockNavigate).toHaveBeenCalledWith('?q=batman')
  })
})
