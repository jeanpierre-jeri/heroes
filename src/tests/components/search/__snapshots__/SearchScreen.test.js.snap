// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1
    className="mt-5"
  >
    Buscar Héroe
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="search"
          >
            Buscar
          </label>
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Buscar un héroe..."
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar a batman y el input con el valor del query string 1`] = `
Array [
  <h1
    className="mt-5"
  >
    Buscar Héroe
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="search"
          >
            Buscar
          </label>
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Buscar un héroe..."
            type="text"
            value="batman"
          />
        </div>
        <button
          className="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card h-100"
          style={
            Object {
              "width": "18rem",
            }
          }
        >
          <img
            alt="Batman"
            className="card-img-top"
            src="/assets/dc-batman.jpg"
          />
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              Detective Comics #27
            </p>
            <a
              className="btn btn-primary"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Ver más...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
